File: Exponential Family Searching/Galloping Search/GallopingSearch.py
def binary_search(
        arr,
        left,
        right,
        target
):

    while left <= right:

        mid = (
            left +
            (right - left) // 2
        )

        if arr[mid] == target:
            return mid

        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1


def galloping_search(
        arr,
        target
):

    if not arr:
        return -1

    if arr[0] == target:
        return 0

    bound = 1

    while (
        bound < len(arr)
        and
        arr[bound] < target
    ):
        bound *= 2

    return binary_search(
        arr,
        bound // 2,
        min(
            bound,
            len(arr) - 1
        ),
        target
    )


arr = [
    5,10,15,20,
    25,30,35,
    40,45,50
]

target = 35

result = galloping_search(
    arr,
    target
)

if result != -1:
    print(
        "Found at index",
        result
    )
else:
    print(
        "Not Found"
    )

File: Exponential Family Searching/Galloping Search/GallopingSearch.java
public class GallopingSearch {

    static int binarySearch(
            int[] arr,
            int left,
            int right,
            int target
    ) {

        while (left <= right) {

            int mid =
                left +
                (right - left) / 2;

            if (arr[mid] == target)
                return mid;

            if (arr[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }

        return -1;
    }

    static int gallopingSearch(
            int[] arr,
            int target
    ) {

        if (arr.length == 0)
            return -1;

        if (arr[0] == target)
            return 0;

        int bound = 1;

        while (
            bound < arr.length &&
            arr[bound] < target
        ) {
            bound *= 2;
        }

        return binarySearch(
            arr,
            bound / 2,
            Math.min(
                bound,
                arr.length - 1
            ),
            target
        );
    }

    public static void main(
            String[] args
    ) {

        int[] arr = {
            5,10,15,20,
            25,30,35,
            40,45,50
        };

        int target = 35;

        int result =
            gallopingSearch(
                arr,
                target
            );

        if (result != -1)
            System.out.println(
                "Found at index "
                + result
            );
        else
            System.out.println(
                "Not Found"
            );
    }
}
